vllm/v1/worker/gpu_worker.py

```diff
@@ -308,15 +312,32 @@ def execute_model(
 
         output = self.model_runner.execute_model(scheduler_output,
                                                  intermediate_tensors)
+
         parallel_config = self.vllm_config.parallel_config
         if parallel_config.distributed_executor_backend != "external_launcher" \
             and not get_pp_group().is_last_rank:
             assert isinstance(output, IntermediateTensors)
             get_pp_group().send_tensor_dict(output.tensors,
                                             all_gather_group=get_tp_group())
-            return None
+            output = EMPTY_MODEL_RUNNER_OUTPUT
+
         assert isinstance(output, ModelRunnerOutput)
-        return output if self.is_driver_worker else None
+        if is_v1_kv_transfer_group():
+            finished_sending, finished_recving = (
+                get_kv_transfer_group().get_finished(
+                    scheduler_output.finished_req_ids))
+            if finished_sending or finished_recving:
+                if output is EMPTY_MODEL_RUNNER_OUTPUT:
+                    output = copy.copy(EMPTY_MODEL_RUNNER_OUTPUT)
+                output.finished_sending = finished_sending
+                output.finished_recving = finished_recving
+
+        if has_kv_transfer_group() or self.is_driver_worker:
+            # with a connector, the scheduler expects output from all workers
+            # otherwise, just from the driver worker
+            return output
+        else:
+            return None
```

Comment:
I don't think we need to worry about the v1 distinction here given this path is v1-only and we are in the process of ripping out the v0 parts of the code anyhow.

```suggestion
        if has_kv_transfer_group():
            finished_sending, finished_recving = (
                get_kv_transfer_group().get_finished(
                    scheduler_output.finished_req_ids))
            if finished_sending or finished_recving:
                if output is EMPTY_MODEL_RUNNER_OUTPUT:
                    output = copy.copy(EMPTY_MODEL_RUNNER_OUTPUT)
                output.finished_sending = finished_sending
                output.finished_recving = finished_recving
            # with a connector, the scheduler expects output from all workers
            return output

        # return output only from the driver worker
        return output if self.is_driver_worker else None
```

Reply:
@orozery sorry, I just saw the earlier [comment](https://github.com/vllm-project/vllm/pull/19555/commits/8143e76d5a3d09aad151a9232e5bc815b3e80908#r2147554837) saying the opposite! 

In any case, I think my comment still stands (and v0 has been deprecated now in any case), thanks for making the change!